packages/node/src/policy-gen/policy-gen-compartment-class.js

```diff
@@ -163,48 +173,63 @@ const makeGetOverriddenResourceNames = (compartmentMap, resources) => {
         return
       }
 
-      /* c8 ignore next */
-      if (compartmentMap.entry.compartment === compartmentName) {
-        // This also "should never happen", as a foreign compartment should
-        // not be importing the entry compartment. Also, the entry
-        // compartment cannot have a canonical name, and we need a canonical
-        // name to apply policy
-        throw new ReferenceError(
-          `Unexpected entry compartment encountered in ${moduleDescriptorCompartment.label}; this is a bug`
-        )
-      }
-
       return moduleDescriptorCompartment
     }
+    const canonicalName = getCanonicalName(compartmentDescriptor)
+    const packagePolicy =
+      policyOverride.resources[canonicalName]?.packages ?? {}
 
     /**
-     * A set of keys from {@link LavaMoatPolicyOverrides.resources} matching
+     * A set of keys from {@link LavaMoatPolicy.resources} matching
      * {@link CompartmentDescriptor.modules the current `CompartmentDescriptor`'s `modules`}.
      * These will be updated by {@link updateModuleSource}.
      *
      * Cached on {@link compartmentDescriptor}.
-     *
-     * @type {Set<string>}
      */
     const overriddenResources = entries(compartmentDescriptor.modules).reduce(
       (
         overriddenResources,
         [moduleDescriptorName, { compartment: moduleDescriptorCompartmentName }]
       ) => {
+        if (!moduleDescriptorCompartmentName) {
+          return overriddenResources
+        }
         const otherCompartmentDescriptor = getValidCompartmentDescriptor(
           moduleDescriptorCompartmentName
         )
-        if (!otherCompartmentDescriptor) {
+        if (
+          !otherCompartmentDescriptor ||
+          otherCompartmentDescriptor === compartmentDescriptor
+        ) {
           return overriddenResources
         }
-        const canonicalName = getCanonicalName(otherCompartmentDescriptor)
-        if (hasValue(resources, canonicalName)) {
-          overriddenResources.add(moduleDescriptorName)
+
+        const otherCanonicalName = getCanonicalName(otherCompartmentDescriptor)
+
+        if (!(otherCanonicalName in packagePolicy)) {
+          return overriddenResources
         }
+
+        // this is _not_ the canonical name, but the compartment.name field,
+        // which is kinda-sorta like a package name
+        overriddenResources.add(moduleDescriptorName)
+
         return overriddenResources
       },
-      new Set()
+      /** @type {Set<string>} */ (new Set())
     )
+
+    for (const [otherCanonicalName, otherPackagePolicyEntry] of entries(
+      packagePolicy
+    )) {
```

Comment:
I'm not sure how effective this is. compartmentDescriptor.modules should contain everything the compartment will be able to import anyway, so anything that didn't match in the `entries(compartmentDescriptor.modules).reduce` above should not work anyway. 
If that's not the case, the for loop should be enough to cover everything. 
Some of the items that are rejected in the loop above are either rejected needlessly or re-added by the for loop.